Render Video

dmclain opened the pull request:
Provide an easy way to render the embed code for EmbeddedVideoFields

Commit:
convert youtube backend to use templates

Commit:
convert vimeo to layout based rendering

Commit:
use patched context so we don't explode the settings file

Commit:
render_video template filter for rendering the new templates

tswicegood commented:
This looks good, but would look better with changes to the `./docs/video.rst` to show how you can create a new backend by subclassing `BaseVideoBackend`.  Would also be worth starting the docs for the new `layout` kwarg.

niran commented on armstrong/core/arm_content/templatetags/content_helpers.py:
```diff
@@ -7,3 +8,8 @@
     return get_preset_thumbnail(value, arg).url
+
+@register.filter
+def render_video(value, arg):
```
Isn't this eventually going to take arguments? (E.g. `layout=nondefault`) If so, changing it to a `simple_tag` seems more future-proof. Filter arguments suck.

tswicegood replied:
-1 on `simple_tag`.  I'd much rather see:

```
if arg.find("=") is -1:
    arg = "layout=%s" % arg
args = urlparse.parse_qs(arg)
return value.backend.embed(value, **args)
```

That supports both of these syntaxes inside the template:

```
{{ article.video|render_video:"huge" }}
{{ article.video|render_video:"layout=big&width=120" }}
```

dmclain replied:
@niran, currently if you wanted to specify `layout=nondefault` you'd do it in settings with a ARMSTRONG_PRESET like::

```
'qvga': {'width': 320, 'height': 240, 'layout': 'nondefault'}
```

dmclain replied:
@tswicegood, you're incorrect with regards to how your code works, passing "huge" doesn't choose the huge layout it chooses the huge preset and renders the default layout. The huge preset could render a different layout if that's what was specified

tswicegood replied:
Nice use of `qvga`.

I guess I'm ok with that.  I want to keep this a filter regardless.

tswicegood replied:
@dmclain re-correctness -- I know it changes it.  I still think we're going to end up with a lot of `nondefault` values that end up getting declared twice rather than just once in the templates.  I think display related code belongs in the templates, but I'll live with this model.

dmclain replied:
i copy/pasted qvga out of the fabfile

niran commented on armstrong/core/arm_content/video/backends/base.py:
```diff
@@ -0,0 +1,17 @@
+        template = get_template('embedded_video/%s/%s.html' % (self.type, layout))
+        kwargs['embed'] = embed
+        kwargs['width'] = width
+        kwargs['height'] = height
```
At some point, we'll need to support specifying the width only, then calculating the height based on the video's aspect ratio.

tswicegood replied:
That can/should be done inside `inject_defaults`.  It's probably worth adding a TODO for it.

tswicegood commented on armstrong/core/arm_content/templatetags/content_helpers.py:
```diff
@@ -1,5 +1,6 @@
 from django import template
 from armstrong.core.arm_content.images.sorl import get_preset_thumbnail
+from armstrong.core.arm_content.images.presets import get_preset_args
```
This should be a relative import, as should the one above it.

Commit:
allow for passing render_video args manually

Commit:
type should be lower case for betterlooking pathnames

Commit:
use folder over type, type is too overloaded and has test cases testing for YouTube, which looks bad as a path

Commit:
use a better parsing method for render_video and add a test case for it

niran commented on armstrong/core/arm_content/video/backends/base.py:
```diff
@@ -0,0 +1,17 @@
+class BaseVideoBackend(object):
+    folder = 'base'
+    
+    @inject_defaults
```
While we're making these changes, I think it's worth ripping out `inject_defaults`. Templates make it unnecessary, and presets doubly so.

Commit:
simplify embed code

Commit:
add comment to explain slightly odd parse_qsl usage

Commit:
add __len__ to EmbeddedVideo to make it pass validators properly

Commit:
Remove __len__, it was a bad idea

Commit:
Remove validators from EmbeddedVideoField

The validators on the form field will work on the raw url, but the DB field is trying to validate an EmbeddedVideo object which will not work

Commit:
remove @inject_defaults from default implementation so that it can be possible to only specify width or only height

joncotton commented:
This is good work, but it's time has past. Sorry Dave. The video code is going to be deprecated. All of it is more flexibly handled in the standalone [armstrong.apps.embeds](https://github.com/armstrong/armstrong.apps.embeds).
